src/ReverseProxy/WebSocketsTelemetry/WebSocketsTelemetryMiddleware.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Yarp.ReverseProxy.WebSocketsTelemetry
+{
+    internal sealed class WebSocketsTelemetryMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        public WebSocketsTelemetryMiddleware(RequestDelegate next)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+        }
+
+        public Task Invoke(HttpContext context)
+        {
+            if (WebSocketsTelemetry.Log.IsEnabled())
+            {
+                if (context.Features.Get<IHttpUpgradeFeature>() is { } upgradeFeature)
+                {
+                    return InvokeAsyncCore(context, upgradeFeature, _next);
+                }
+            }
+
+            return _next(context);
+        }
+
+        private static async Task InvokeAsyncCore(HttpContext context, IHttpUpgradeFeature upgradeFeature, RequestDelegate next)
+        {
+            var upgradeWrapper = new HttpUpgradeFeatureWrapper(upgradeFeature);
+            context.Features.Set<IHttpUpgradeFeature>(upgradeWrapper);
+
+            try
+            {
+                await next(context);
+            }
+            finally
+            {
+                if (upgradeWrapper.TelemetryStream is { } telemetryStream)
+                {
+                    WebSocketsTelemetry.Log.WebSocketClosed(
+                        telemetryStream.EstablishedTime.Ticks,
+                        telemetryStream.ClientCloseTime?.Ticks ?? -1,
+                        telemetryStream.ServerCloseTime?.Ticks ?? -1,
+                        telemetryStream.MessagesRead,
+                        telemetryStream.MessagesWritten);
+                }
```

Comment:
Best practice: Unwrap/cleanup on exit
```suggestion
                }
                context.Features.Set<IHttpUpgradeFeature>(upgradeFeature);
```